sobel/main.c

```diff
@@ -0,0 +1,143 @@
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <pthread.h>
+
+extern void* sobel(void* args);
+
+// struct for transporting data between threads
+struct data {
+    int width;
+    int start;
+    int end;
+    char** base;
+    char** res;
+};
+
+int main(int argc, char* argv[]) {
+    if (argc != 2) {
+        printf("No input file or too much arguments\n");
+        exit(1); 
+    }
+    
+    char* filename = argv[1];
+    FILE* input;
+    FILE* output;
+    
+    // checking the file
+    if ((input = fopen(filename, "r")) == NULL) {
+        printf("Wrong filename");
+        exit(1);
+    }
+    
+    char* info = (char*)malloc(54);
+    
+    // reading .bmp info
+    for (int i = 0; i < 54; ++i) {
+        info[i] = fgetc(input);
+    }
+    
+    int width = *(int*)&info[18];
+    int height = *(int*)&info[22];
+    char r, g, b;
+    int x;
+    
+    // allocating memory
+    char** base = (char**)malloc(height * sizeof(char*));
+    char** res = (char**)malloc(height * sizeof(char*));
+    for (int i = 0; i < height; ++i) {
+        base[i] = (char*)malloc(width);
+        res[i] = (char*)malloc(width);
+    }
+    
+    for (int i = 0; i < height; ++i) {
+        for (int j = 0; j < width; ++j) {
+            // reading red, green and blue bytes and getting one grayscale byte
+            r = fgetc(input);
+            g = fgetc(input);
+            b = fgetc(input);
+            x = (int)(0.299 * r + 0.587 * g + 0.114 * b) / 3;
```

Comment:
Please, avoid using magic numbers in code, move them to global const variables or defines. https://en.wikipedia.org/wiki/Magic_number_(programming)